File: 1-working-with-list/p07_flatten.ml

type 'a node =
  | One of 'a
  | Many of 'a node list;;

let rec flatten = function
  | [] -> []
  | hd :: tl ->
    match hd with
    | One e -> e :: flatten tl
    | Many l -> flatten l @ flatten tl;;

let flatten' l =
  let rec call acc = function
    | [] -> acc
    | hd :: tl ->
      match hd with
      | One e -> call (acc @ [e]) tl
      | Many l -> call (call acc l) tl
  in
  call [] l;;

let flatten'' l =
  let rec call acc = function
    | [] -> acc
    | One e :: tl -> call (e::acc) tl
    | Many l :: tl -> call (call acc l) tl
  in
  List.rev (call [] l);;

flatten [ ];;
flatten [ One "a" ];;
flatten [ Many [ One "b" ; Many [ One "c" ; One "d" ] ] ];;
flatten [ One "a" ; Many [ One "b" ; Many [ One "c" ; One "d" ] ; One "e" ] ];;

flatten' [ ];;
flatten' [ One "a" ];;
flatten' [ Many [ One "b" ; Many [ One "c" ; One "d" ] ] ];;
flatten' [ One "a" ; Many [ One "b" ; Many [ One "c" ; One "d" ] ; One "e" ] ];;

flatten'' [ ];;
flatten'' [ One "a" ];;
flatten'' [ Many [ One "b" ; Many [ One "c" ; One "d" ] ] ];;
flatten'' [ One "a" ; Many [ One "b" ; Many [ One "c" ; One "d" ] ; One "e" ] ];;

File: 1-working-with-list/p07_flatten.scala

object Main {

  sealed abstract class Node[T]
  case class One[T](t: T) extends Node[T]
  case class Many[T](l: List[Node[T]]) extends Node[T]

  def main(args: Array[String]) {
    val l = List(One("a"), Many(List(One("b"), Many(List(One("c"), One("d"))), One("e"))))
    println(flatten(l))
  }

  def flatten[T](l: List[Node[T]]): List[T] = {
    l match {
      case Nil => Nil
      case hd :: rest =>
        hd match {
          case One(t) => t :: flatten(rest)
          case Many(t) => flatten(t) ::: flatten(rest)
        }
    }
  }

}
